[victronenergydbus][victronenergyvrm] Initial contributions

SamLue opened the pull request:
<!-- TITLE -->

- "[victronenergydbus] Initial contribution"
- "[victronenergyvrm] Initial contribution"

<!-- DESCRIPTION -->

I developed two new bindings for monitoring SolarCharger from Victron Energy.
One for directly connected SolarCharger via USB (victronenergydbus) and the other for getting the data from api https://vrmapi.victronenergy.com/ 
Any Things and Channels are described at README.md

This is my first PR. I hope my misstakes aren't to big.

Hilbrand commented on addons/binding/org.openhab.binding.victronenergydbus/src/main/java/org/openhab/binding/victronenergydbus/internal/VictronEnergyDBusSolarCharger.java:
```diff
@@ -0,0 +1,257 @@
+
+    public String getStateStr() {
+        int state = getState();
+        if (state == 0) {
```
Use a `switch` statement.

Hilbrand commented on addons/binding/org.openhab.binding.victronenergydbus/src/main/java/org/openhab/binding/victronenergydbus/internal/VictronEnergyDBusSolarCharger.java:
```diff
@@ -0,0 +1,257 @@
+
+    public String getErrString() {
+        int error = getErr();
+        if (error == 0) {
```
Use a `switch` statement.

Hilbrand commented on addons/binding/org.openhab.binding.victronenergyvrm/src/main/java/org/openhab/binding/victronenergyvrm/handler/VictronEnergyVRMHandler.java:
```diff
@@ -0,0 +1,146 @@
+            updateStatus(ThingStatus.ONLINE);
+            logger.debug("Login successfully");
+            // logger.debug("Here is the token: " + token);
+            pollingJob = scheduler.scheduleWithFixedDelay(this::updateData, 0, DEFAULT_REFRESH_RATE, TimeUnit.SECONDS);
```
You need to stop this polling job in the `dispose` method.

Hilbrand commented on addons/binding/org.openhab.binding.victronenergyvrm/src/main/java/org/openhab/binding/victronenergyvrm/handler/VictronEnergyVRMHandler.java:
```diff
@@ -0,0 +1,146 @@
+            logger.debug("Increment Errorcounter.");
+            errorcounter++;
+            if (errorcounter < 3) {
+                this.initialize();
```
This call will reassign the `pollingJob` in `initialize` to a new scheduled job and leak the previous assigned scheduled job, adding a new scheduled job each time this code is called.

SamLue commented:
Hello Hilbrand,

thanks for your comments.

But as you write "...changes might look big, ..." I won't agree with all of your aspects and can't do/change all of your points.

In my point of view it make sense to create 2 bindings. The difference is the way to communicate with the device(s). In one case (DBus) the device is locally connected (via USB). In the other case (VRM) the binding gets all data from an api. So I will use these two bindings in complete different scenarios. If I look out of the User-Glasses, it would be more clear and I would understand faster in which case I have to install/configure which things.

It was a really big task to get the dbus communication running. I changed the code multiple times, had contact with a developer of dbus-java library, and updated everything to use the latest lib java-dbus 3.0.0. Besides that I don't have any information about how to use the transport bundle, I won't change it to an older way/library only to be "openhab-default".

Ok, I agree with the http-client. It is recommended to use jetty here: [https://www.eclipse.org/smarthome/documentation/development/bindings/dependencies.html], so my fault that I overread this.

All the other things we can talk about. No problem.

So, there are two options now. The first is, you say it's ok to release two different bindings and the implementation of dbus library can be as it is. Then I will begin with all the other work. But if you insist on all also the first two points, I won't do anything.

For me my bindings are working fine at the moment and i don't need them in the default add-ons. I can install it manually. I thought about the community to share my work by helping with a very small piece of the puzzle to growth up the community.

So now it's your choice.

Best regards,

Samuel

J-N-K commented:
@Hilbrand, after looking in the documentation the things look quite different, so I tend to agree with @SamLue. Can you comment?

J-N-K commented:
@SamLue Do you intend to continue on this one? 
@Hilbrand Can you share your thoughts`?

SamLue commented:
@J-N-K I already have changed the http-client in [victronenergyvrm] to jetty. But only in my and some other interested users installation. I havn't done a pull request for these changes yet.
Can you tell me what the nessesary tasks are to get merged?

cpmeister commented:
You don't need to make a new PR to get your http-client changes merged. You can simply make additional commits to the branch that his PR is from (in this case SamLue:master). Please add additional commits as necessary as you address issues we bring up. Thanks!

cpmeister commented:
@SamLue 
It looks like all of your files are located under `addons/bindings/` directory. That isn't where bindings below. They should be located under the 'bundle` directory. Please move all your files to the correct directory.
